src/twisted/conch/ssh/factory.py

```diff
@@ -56,7 +56,13 @@ def buildProtocol(self, addr):
         @return: The built transport.
         """
         t = protocol.Factory.buildProtocol(self, addr)
-        t.supportedPublicKeys = self.privateKeys.keys()
+        # Advertise only the host-key algorithms that correspond to
+        # keys we have
```

Comment:
```suggestion
        # Advertise only the host-key algorithms that correspond to
        # the intersection of what we have with what we support.
```

I don't know what to say about this change.... but I am not against it :)

This will silently ignore a "configured" private key when python cryptography/openssl does not support it or for example if you accidentally add  a private key with a typo in the name (ssh-dsa vs ssh-dss :) 

I guess that generating a log when a non-supported key is configured can help detect some configuration / deployment errors.

I think that converting this to a list here is OK and adding the log should not be a blocker.

-------------

Reply:
I think this does not significantly change the behavior if a private key is misconfigured — we would never have been able to use that key, since our connection peer wouldn't have negotiated for `ssh-dsa`.

(Converting to a list is the main motivation for this PR as I understand it. It was a list on Python2, but on Python3 it turned into a dict keys view, which doesn't support indexing. The list-ification just restores the status-quo-ante.)